### PR TITLE
feat: Add flagz endpoint for kube-scheduler

### LINKED ISSUE #128816: Add flagz endpoint for kube-scheduler
### What would you like to be added?

Part of 
- https://github.com/kubernetes/enhancements/issues/4828

Add the `/flagz` endpoint for kube-scheduler.

Sample response:
```
----------------------------
title: Kubernetes Flagz
description: Command line flags that Kubernetes component was started with.
----------------------------

default-watch-cache-size=100
delete-collection-workers=1
enable-garbage-collector=true
encryption-provider-config-automatic-reload=false
...
```

/sig instrumentation

### Why is this needed?

Refer to [Motivation setcion](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/4828-component-flagz/README.md#motivation) in KEP.

### PR DESCRIPTION
**Pull request by yongruilin**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Adds a /flagz endpoint for kube-scheduler

#### Which issue(s) this PR fixes:

Fixes #128816 and https://github.com/kubernetes/enhancements/issues/4828

#### Does this PR introduce a user-facing change?

```release-note
Adds a /flagz endpoint for kube-scheduler endpoint
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
* KEP : https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/4828-component-flagz/README.md

Example response
```
~ curl -k --cert /etc/kubernetes/pki/apiserver-kubelet-client.crt --key /etc/kubernetes/pki/apiserver-kubelet-client.key https://localhost:10259/flagz
```
```
kube-scheduler flags
Warning: This endpoint is not meant to be machine parseable, has no formatting compatibility guarantees and is for debugging purposes only.

allow-metric-labels []
allow-metric-labels-manifest
authentication-kubeconfig /etc/kubernetes/scheduler.conf
authentication-skip-lookup false
authentication-token-webhook-cache-ttl 10s
authentication-tolerate-lookup-failure true
authorization-always-allow-paths [/healthz,/readyz,/livez]
authorization-kubeconfig /etc/kubernetes/scheduler.conf
authorization-webhook-cache-authorized-ttl 10s
authorization-webhook-cache-unauthorized-ttl 10s
bind-address 127.0.0.1
cert-dir
client-ca-file
config
contention-profiling true
disable-http2-serving false
disabled-metrics []
emulated-version []
feature-gates :ComponentFlagz=true
help false
http2-max-streams-per-connection 0
kube-api-burst 100
kube-api-content-type application/vnd.kubernetes.protobuf
kube-api-qps 50
kubeconfig /etc/kubernetes/scheduler.conf
leader-elect true
leader-elect-lease-duration 15s
leader-elect-renew-deadline 10s
leader-elect-resource-lock leases
leader-elect-resource-name kube-scheduler
leader-elect-resource-namespace kube-system
leader-elect-retry-period 2s
log-flush-frequency 5s
log-json-info-buffer-size 0
log-json-split-stream false
log-text-info-buffer-size 0
log-text-split-stream false
logging-format text
master
permit-address-sharing false
permit-port-sharing false
pod-max-in-unschedulable-pods-duration 5m0s
profiling true
requestheader-allowed-names []
requestheader-client-ca-file
requestheader-extra-headers-prefix [x-remote-extra-]
requestheader-group-headers [x-remote-group]
requestheader-uid-headers []
requestheader-username-headers [x-remote-user]
secure-port 10259
show-hidden-metrics-for-version
tls-cert-file
tls-cipher-suites []
tls-min-version
tls-private-key-file
tls-sni-cert-key []
v 0
version false
vmodule
write-config-to
```

```docs
[KEP] https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/4828-component-flagz/README.md
```

### CODE REVIEW

File: cmd/kube-scheduler/app/options/options.go [hunk 1]
```diff
@@ -330,6 +331,9 @@ func (o *Options) Config(ctx context.Context) (*schedulerappconfig.Config, error
 	dynClient := dynamic.NewForConfigOrDie(c.KubeConfig)
 	c.DynInformerFactory = dynamicinformer.NewFilteredDynamicSharedInformerFactory(dynClient, 0, corev1.NamespaceAll, nil)
 	c.LeaderElection = leaderElectionConfig
+	if o.Flags != nil {
```
Comment (haosdent) [inline cmd/kube-scheduler/app/options/options.go:334]: Should we add
```
if utilfeature.DefaultFeatureGate.Enabled(zpagesfeatures.ComponentFlagz) {
```
at here?

File: cmd/kube-scheduler/app/server.go [hunk 2]
```diff
@@ -68,6 +70,10 @@ import (
 	"k8s.io/kubernetes/pkg/scheduler/profile"
 )
 
+const (
+	kubeScheduler = "kube-scheduler"
```
Comment (kerthcet) [inline cmd/kube-scheduler/app/server.go:74]: Can you replace L94 with this variable, other places as well. Thanks.

File: cmd/kube-scheduler/app/options/options.go [hunk 3]
```diff
@@ -330,6 +332,11 @@ func (o *Options) Config(ctx context.Context) (*schedulerappconfig.Config, error
 	dynClient := dynamic.NewForConfigOrDie(c.KubeConfig)
 	c.DynInformerFactory = dynamicinformer.NewFilteredDynamicSharedInformerFactory(dynClient, 0, corev1.NamespaceAll, nil)
 	c.LeaderElection = leaderElectionConfig
+	if utilfeature.DefaultFeatureGate.Enabled(zpagesfeatures.ComponentFlagz) {
```
Comment (kerthcet) [inline cmd/kube-scheduler/app/options/options.go:335]: Can we move to function ApplyTo?

File: cmd/kube-scheduler/app/server.go [hunk 4]
```diff
@@ -238,7 +244,15 @@ func Run(ctx context.Context, cc *schedulerserverconfig.CompletedConfig, sched *
 
 	// Start up the healthz server.
 	if cc.SecureServing != nil {
-		handler := buildHandlerChain(newHealthEndpointsAndMetricsHandler(&cc.ComponentConfig, cc.InformerFactory, isLeader, checks, readyzChecks), cc.Authentication.Authenticator, cc.Authorization.Authorizer)
+		unsecuredMux := newHealthEndpointsAndMetricsHandler(&cc.ComponentConfig, cc.InformerFactory, isLeader, checks, readyzChecks)
+
+		if utilfeature.DefaultFeatureGate.Enabled(zpagesfeatures.ComponentFlagz) {
```
Comment (kerthcet) [inline cmd/kube-scheduler/app/server.go:249]: Can we move this to the `newHealthEndpointsAndMetricsHandler`, then no need to change the return type. We can rename the function to like `newEndpointsHandler`.

**Review by kerthcet — COMMENTED** [review]
We need a test to make sure flagz works, see similar ones here: test/integration/scheduler/serving/healthcheck_test.go

File: staging/src/k8s.io/component-base/zpages/flagz/flagreader.go [hunk 5]
```diff
@@ -25,7 +25,7 @@ type Reader interface {
 	GetFlagz() map[string]string
 }
 
-// NamedFlagSetsGetter implements Reader for cliflag.NamedFlagSets
+// NamedFlagSetsReader implements Reader for cliflag.NamedFlagSets
```
Comment (kerthcet) [inline staging/src/k8s.io/component-base/zpages/flagz/flagreader.go:28]: Good cache, thanks

File: cmd/kube-scheduler/app/server.go (hunk 4 above)
Comment (yongruilin) [reply cmd/kube-scheduler/app/server.go]: Sounds good. Updated.

**Comment by kerthcet** [comment]
> We need a test to make sure flagz works, see similar ones here: test/integration/scheduler/serving/healthcheck_test.go

We need one more test to make sure it works. @yongruilin Thanks

**Comment by yongruilin** [comment]
@kerthcet thanks for the review, I've added the integration test for flagz endpoint. Follow the same pattern in https://github.com/kubernetes/kubernetes/pull/128848

File: test/integration/scheduler/serving/flagz_test.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+	"testing"
+
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
+	"k8s.io/component-base/zpages/features"
+	"k8s.io/klog/v2/ktesting"
+	kubeschedulertesting "k8s.io/kubernetes/cmd/kube-scheduler/app/testing"
+)
+
+func TestFlagzEndpoints(t *testing.T) {
```
Comment (kerthcet) [inline test/integration/scheduler/serving/flagz_test.go:33]: Can we reuse the `TestHealthEndpoints` and rename it to `TestEndpoints`, table driven test is more welcomed in kubernetes codebase, and I think it's feasible by just adding more cases there? WDYT?

Comment (yongruilin) [reply test/integration/scheduler/serving/flagz_test.go]: Updated, PTAL again. I fit the /flagz into the original test table.

File: cmd/kube-scheduler/app/server.go [hunk 7]
```diff
@@ -238,7 +244,15 @@ func Run(ctx context.Context, cc *schedulerserverconfig.CompletedConfig, sched *
 
 	// Start up the healthz server.
```
[the remark this answers is not in the document]
Comment (yongruilin) [reply cmd/kube-scheduler/app/server.go]: Updated.

**Review by kerthcet — COMMENTED** [review]
Leave the final LGTM to @dom4ha  and please squash, thanks for your patience and efforts.

/approve

**Comment by dom4ha** [comment]
/lgtm
It looks great now, thanks.

**Pull request merged**